File: Difficulty: Medium/0 - 1 Knapsack Problem/0-1-knapsack-problem.js
//{ Driver Code Starts
// Initial Template for JavaScript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split('\n').map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    let testCases = parseInt(readLine());
    for (let i = 0; i < testCases; i++) {
        let inputLine = readLine().split(' ');
        let capacity = parseInt(inputLine[0]); // Capacity of the knapsack

        let values = readLine().split(' ').map(x => parseInt(x)); // Values of the items
        let weights =
            readLine().split(' ').map(x => parseInt(x)); // Weights of the items

        let solution = new Solution();
        let maxValue = solution.knapSack(capacity, values, weights); // Maximum value
        if (maxValue === -0) maxValue = 0; // Handle potential -0
        console.log(maxValue);
        console.log("~");
    }
}

// } Driver Code Ends


// User function Template for javascript

/**
 * @param {number} capacity
 * @param {number[]} val
 * @param {number[]} wt
 * @returns {number}
 */

class Solution {
    // Function to return max value that can be put in knapsack of capacity.
    knapSack(capacity, val, wt) {
        function f(capacity, index) {
        if (index >= val.length || capacity < 0)
            return 0;
        let include = 0;
        if (capacity - wt[index] >= 0)
            include = val[index] + f(capacity - wt[index], index + 1);
        let exclude = f(capacity, index + 1);
        return Math.max(include, exclude);
    }
    return f(capacity, 0)
    }
}

File: Difficulty: Medium/0 - 1 Knapsack Problem/0-1-knapsack-problem.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    // Function to return max value that can be put in knapsack of capacity.
    int f(int capacity, vector<int>& val, vector<int>& wt, int index) {
        if (index >= val.size() || capacity < 0) return 0;
        
        int include = 0;
        if (capacity - wt[index] >= 0)
            include = val[index] + f(capacity - wt[index], val, wt, index + 1);
        int exclude = f(capacity, val, wt, index + 1);
        return max(include, exclude);
    }
    
    int knapSack(int capacity, vector<int> &val, vector<int> &wt) {
        int ans = f(capacity, val, wt, 0);
        return ans;
    }
};


//{ Driver Code Starts.

int main() {
    // Taking total test cases
    int testCases;
    cin >> testCases;
    cin.ignore();
    while (testCases--) {
        // Reading number of items and capacity
        int numberOfItems, capacity;
        vector<int> weights, values;
        string input;
        int number;

        // Read capacity and number of items
        getline(cin, input);
        stringstream ss(input);
        ss >> capacity;      // The first number is the capacity
        ss >> numberOfItems; // The second number is the number of items

        // Read values
        getline(cin, input);
        ss.clear();
        ss.str(input);
        while (ss >> number) {
            values.push_back(number);
        }

        // Read weights
        getline(cin, input);
        ss.clear();
        ss.str(input);
        while (ss >> number) {
            weights.push_back(number);
        }

        Solution solution;
        cout << solution.knapSack(capacity, values, weights) << endl;
        cout << "~" << endl;
    }
    return 0;
}

// } Driver Code Ends